### DISCUSSION TITLE
Using 5GHz WiFi on a MicroPython board

### DISCUSSION THREAD

**Discussion by piotrwest** [micropython/micropython#11609] [Core Development]
Hi folks! I wanted to use MicroPython on a "rich" board (in terms of I/O & ADCs) with 5GHz WiFi. Unfortunately, the number of chips supporting 5GHz WiFi is pretty limited. It looks like the most popular and affordable module supporting 5GHz WiFi is RTL8720.

RTL8720 is a pretty simple board. It would be great to connect it, for example, with pyboard to provide 5GHz connectivity.

I was looking for ways to achieve such integration. I'm considering those options:
1. Use UART and communicate with RTL8720 running official AT firmware. Send AT Commands.
 - cons: no library available to access WiFi over UART in MicroPython "native" way on pyboard side.
 - pros: RTL8720 runs official AT firmware, relatively tested.
2. Use I2C/SPI/UART to communicate with RTL8720 running custom firmware. Send custom commands.
 - cons: same as before - no library available to access WiFi over I2C/SPI/UART in MicroPython "native" way.
 - pros: RTL8720 could run MicroPython: [ambiot/micropython/releases/tag/v1.1.0-ameba](https://github.com/ambiot/micropython/releases/tag/v1.1.0-ameba) - supposedly wifi is working: [amebad-micropython-bw16-getting-started](https://www.amebaiot.com/en/amebad-micropython-bw16-getting-started/) - making the integration a bit faster in terms of development effort.
3. Use eRPC to communicate with RTL8720, running [Seeed_Arduino_rpcUnified](https://github.com/Seeed-Studio/Seeed_Arduino_rpcUnified)
 - cons: the Seeed Studio rpcWiFi library would have to be wrapped in MicroPython functions.
 - pros: it looks like the integration is tested - a lot of examples of rpcWiFi on MQTT, HTTPS, UDP: [Wio-Terminal-Wi-Fi](https://wiki.seeedstudio.com/Wio-Terminal-Wi-Fi)

I'm looking for advice - what would you recommend as the least effort and reliable way?
Also, would it be very hard to build an abstract WiFi adapter (e.g. WiFi over UART) in MicroPython, so that libraries using the network interface work effortlessly? Do you have any pointers to source code of interest?

Thanks!

**Comment by rkompass** [comment]
For option 1 perhaps [this Lib](https://github.com/noyelseth/rpi-pico-micropython-esp8266-lib/blob/main/esp8266.py) may be useful.
As option 4 I would suggest to run MP on the RTL8720 and do the 'rich IO' on the STM (Py-) board. There you have tested and reliable known timing behaviour. It might be easier to control the Pyboard than other way round.

**Comment by mattytrentini** [comment]
There is an [Ameba port](https://github.com/ambiot/micropython/tree/ameba) in development - so you could run MicroPython _on_ the RTL8720. The port is very much a WIP but Realtek employees are working on it, there is [preliminary support](https://www.amebaiot.com/en/ameba-micropython-summary/), and they would be keen to get feedback on it! There's also an Ameba channel in the MicroPython Discord, though it's quite new.

For what it's worth, I've run that MicroPython port on my AMB21 and AMB23 and connected them to a 5GHz network. 

([Ameba](https://www.amebaiot.com/en/) is a division of Realtek that covers their IoT devices, including the RTL8720.)

**Reply by piotrwest** [reply]
Thanks! Do you know if the final intent is to merge in the RTL8720 to official MicroPython repo?

**Reply by mattytrentini** [reply]
That was the _intent_ - but there was a lot of work to be done...

**Reply by mvdw73** [reply]
I was originally looking to work on this a couple of years ago, but work and life got in the way. Now I have some time and also a pressing need for a similar requirement as yourself: I need a board with 5GHz wifi, on which I can run something nicely embedded. Micropython is a Nice to Have, and with the Ameba port being reasonably complete, I am going to try to fix the issues I'm having with running an MQTT client on the board.

If you don't need to run MQTT, and very simple sockets will do your job, it works quite nicely.

I have some BW16 boards I've managed to load with micropython, and opening a socket to a linux computer running netcat allows me to send arbitrary data over the 5GHz wifi. 

On the linux computer:

```bash
nc -l -p 2000
```

On the BW16 micropython board:

```python
from wireless import WLAN
import socket

ssid = "SSID"
pwd = "Password"

wifi = WLAN(mode=WLAN.STA)
wifi.connect(ssid=ssid, pswd = pwd)

s = socket.socket()
s.connect("ip.of.linux.computer", 2000)
s.send("Hello from Micropython!\n")

```

And you should see `Hello from Micropython!` on your linux terminal.

**Comment by piotrwest** [comment]
Thanks folks. It looks like one could write a local proxy, listen to incoming bytes and forward everything through AT commands to the target. This way you get a file descriptor on the input and you could use such solution with ussl. It's questionable though if this would be more reliable than just using RTL8720 as primary uC... I'm not quite sure.

**Comment by mvdw73** [comment]
> That was the _intent_ - but there was a lot of work to be done...

What needs to be done to get the ameba port integrated in the upstream micropython tree? Is there a list somewhere of what Damien requires before a port is accepted into the upstream tree?

**Reply by mattytrentini** [reply]
Here are some [relevant comments](https://github.com/micropython/micropython/pull/5449#issuecomment-714454904) from @dpgeorge on the topic. It's difficult; implementing a port is time-consuming and challenging - but it's even harder for the MicroPython folks that ultimately have to take over and support it if it's merged upstream...
